R/install.R

```diff
@@ -1,276 +1,343 @@
+
+  ## Cut stdout and stderr to lines
+  worker$stdout <- cut_into_lines(worker$stdout)
+  worker$stderr <- cut_into_lines(worker$stderr)
+
+  ## Record what was done
+  stop_task(state, worker)
 }
 
+select_next_task <- function(state) {
+
+  ## Cannot run more workers?
+  if (length(state$workers) >= state$config$num_workers) {
+    return(task("idle"))
+  }
+
+  ## Can we select a source package build? Do that.
+  can_build <- which(
+    ! state$plan$build_done &
+    vapply(state$plan$dependencies, length, integer(1)) == 0 &
+    is.na(state$plan$worker_id))
 
-#' @importFrom pkgdepends remotes
-get_install_plan <- function(filenames, library) {
-  r <- remotes$new(glue("local::{filenames}"), library = library)
-  r$solve()
-  r$download_solution()
-  plan <- r$get_install_plan()
+  if (any(can_build)) {
+    pkgidx <- can_build[1]
+    return(task("build", pkgidx = pkgidx))
+  }
 
-  installed <- plan$type == "installed"
+  ## TODO: can we select a binary that is depended on by a source package?
 
-  # Remove installed dependencies from the dependency lists.
-  plan$dependencies <- lapply(plan$dependencies, setdiff, plan$package[installed])
+  ## Otherwise select a binary if there is one
+  can_install <- which(
+    state$plan$build_done &
+    ! state$plan$install_done &
+    is.na(state$plan$worker_id))
 
-  # Return uninstalled packages
-  plan[!installed, ]
-}
+  if (any(can_install)) {
+    pkgidx <- can_install[1]
+    return(task("install", pkgidx = pkgidx))
+  }
 
-get_output <- function(x) {
-  res <- poll(x$processes, -1)
-  idx <- which(vapply(res, function(x) any(x == "ready"), logical(1)))
-  lapply(idx,
-    function(i) {
-      list(
-        output = remove_spaces(x$processes[[i]]$read_output_lines()),
-        error = x$processes[[i]]$read_error_lines(),
-        index = i)
-    })
+  ## Looks like nothing else to do
+  task("idle")
 }
 
-running_processes <- function(x) {
-  length(x$processes) > 0
+task <- function(name, ...) {
+  list(name = name, args = list(...))
 }
 
-kill_all_processes <- function(events) {
-  for (proc in events$processes) {
-    proc$kill(tools::SIGINT)
+start_task <- function(state, task) {
+  if (task$name == "idle") {
+    state
+
+  } else if (task$name == "build") {
+    start_task_build(state, task)
+
+  } else if (task$name == "install") {
+    start_task_install(state, task)
+
+  } else {
+    stop("Unknown task, internal error")
   }
 }
 
-remove_spaces <- function(x) {
-  x <- x[!grepl("^[[:space:]]*$", x)]
-}
+get_worker_id <- (function() {
+  id <- 0
+  function() {
+    id <<- id + 1
+    as.character(id)
+  }
+})()
 
-#' @importFrom crayon make_style strip_style
-greyish <- function() make_style("darkgrey")
+#' @importFrom pkgbuild pkgbuild_process
 
-green_tick <- function() green(symbol$tick)
-red_cross <- function() red(symbol$cross)
+start_task_build <- function(state, task,
+                             dummy = FALSE, dummy_args = list()) {
```

Comment:
I don't really like having external arguments just for testing purposes (even for internal functions), but maybe there is no better way to do this.

Reply:
I'll try to refactor it.

Reply:
Done.